packages/alfa-tuple/src/tuple.ts

```diff
@@ -0,0 +1,390 @@
+import { Array } from "@siteimprove/alfa-array";
+import { Callback } from "@siteimprove/alfa-callback";
+import { Comparer } from "@siteimprove/alfa-comparable";
+import { Hash } from "@siteimprove/alfa-hash";
+import { Mapper } from "@siteimprove/alfa-mapper";
+import { Option } from "@siteimprove/alfa-option";
+import { Predicate } from "@siteimprove/alfa-predicate";
+import { Reducer } from "@siteimprove/alfa-reducer";
+import { Refinement } from "@siteimprove/alfa-refinement";
+
+/**
+ * @public
+ */
+export type Tuple<T extends ReadonlyArray<unknown> = ReadonlyArray<unknown>> =
+  T;
+
+/**
+ * @public
+ */
+export namespace Tuple {
+  export type Item<T extends Tuple> = T[number];
+
+  export function of<T extends Tuple>(...values: T): T {
+    return values;
+  }
+
+  export type Empty = [];
+
+  export function empty(): Empty {
+    return [];
+  }
+
+  export type Size<T extends Tuple> = T["length"];
+
+  export function size<T extends Tuple>(tuple: T): Size<T> {
+    return Array.size(tuple);
+  }
+
+  export type Copy<T extends Tuple> = T extends readonly [...infer T]
+    ? T
+    : Empty;
```

Comment:
Isn't that the same thing as `T`?

Also, I feel it would make sense inside this namespace to use `Tuple<T>` instead of arrays to keep intention clearer:
```typescript
export type Copy<T extends Tuple> = T extends Tuple<[...infer T]>
    ? Tuple<T>
    : Empty;
```
should do the same 🤔 (same remark in various places). (notably if we ever change implementation so that `Tuple<T>` is not anymore a synonym for `T`, these notation wouldn't need to change)

Reply:
It's _usually_ the same as `T`, but will remove any `readonly` modifiers. That's also why `Tuple<T>` won't work as that re-introduces the `readonly` modifiers. Here's how it's intended to work:

```ts
const tuple: readonly [number, null, boolean] = [1, null, true];
const copy: [number, null, boolean] = copy(tuple);
```